### PR TITLE
Add pooling support for osu! hit judgements

### LINKED ISSUE #8250: Lag spikes when displaying animated judgement sprites
**Describe the bug:**  
- Big lag spikes as much as 500ms when the game tries to display an animated judgement sprite (e.g. 100, 50, miss...) 
- Skin tested: [FlyingTuna's skin](https://skins.osuck.net/index.php?newsid=640)
- Most prominent right after game start up, sometimes it goes away after a while sometimes it doesn't
- applies to both gameplay and replay

**Screenshots or videos showing encountered issue:** 
~~will add link after the video finished uploading~~  
https://streamable.com/nejq1

**osu!lazer version:**  
2020.312.0

**Logs:**
[performance.log](https://github.com/ppy/osu/files/4325887/performance.log)
[performance-draw.log](https://github.com/ppy/osu/files/4325889/performance-draw.log)
[performance-update.log](https://github.com/ppy/osu/files/4325890/performance-update.log)

### PR DESCRIPTION
**Pull request by peppy**
Proof of concept mostly for comment / feedback. Only implemented for osu! ruleset initially. ~Note that this doesn't cover preload requirements (will only preload the default result type, which is "Perfect" right now).~

~Further improvement could come from making one pool per result type, avoiding child reconstruction on type change (probably what we want).~ (since added)

- [x] Depends on https://github.com/ppy/osu-framework/pull/3681.
- [x] Add support for non-osu! rulesets (or at least regression test)
- Closes #8250

### CODE REVIEW

**Comment by peppy** [comment]
While I remember, this PR still needs testing and support in non-osu! rulesets. I don't think it should be a huge step to add so feel free to do this during review if it suits (@smoogipoo), else I'll look into it.

**Comment by smoogipoo** [comment]
The amount of work and knowledge required to use the pool is quite... concerning. Everything seems to be done specifically because the framework changes don't provide a way to do it...

1. A dictionary of `HitResult -> DrawablePool` to split them up.
2. A setup action which sets the `DrawableObject` on the `DrawableJudgement`.
    1. Doesn't tell the user how to properly initialise this object for correct use.
    2. It's not explained anywhere unless you actually understand how everything works, but `Result` isn't passed in through this same process. I understand why but if a user uses this, they're more likely to encounter bad behaviour.
3. An override of `CreateNewDrawable` which basically does the same thing as the setup function, except sets the `Result` instead of `DrawableObject`.
4. The `Result` property is backed by a `Cached`, whereas `DrawableObject` isn't. Again this is done for a specific implementation detail of `DrawableJudgement`.
5. Manual asyncification for the preload.

Disregarding (5) because it's not the purpose of the pool, this all seems very DIY which is probably just due to a lack of a strict model->drawable relationship and supporting structure.

I refactored this into a model->drawable relationship and the changes here look something like this:

```diff
diff --git a/osu.Game.Rulesets.Osu.Tests/TestSceneDrawableJudgement.cs b/osu.Game.Rulesets.Osu.Tests/TestSceneDrawableJudgement.cs
index c81edf4e0..fec34e934 100644
--- a/osu.Game.Rulesets.Osu.Tests/TestSceneDrawableJudgement.cs
+++ b/osu.Game.Rulesets.Osu.Tests/TestSceneDrawableJudgement.cs
@@ -19,7 +19,7 @@ public TestSceneDrawableJudgement()
             foreach (HitResult result in Enum.GetValues(typeof(HitResult)).OfType<HitResult>().Skip(1))
             {
                 AddStep("Show " + result.GetDescription(), () => SetContents(() =>
-                    new DrawableOsuJudgement(new JudgementResult(new HitObject(), new Judgement()) { Type = result }, null)
+                    new DrawableOsuJudgement(new DrawableJudgementData(new JudgementResult(new HitObject(), new Judgement()) { Type = result }, null))
                     {
                         Anchor = Anchor.Centre,
                         Origin = Anchor.Centre,
diff --git a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuJudgement.cs b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuJudgement.cs
index 9d0c40629..f7ea00f15 100644
--- a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuJudgement.cs
+++ b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuJudgement.cs
@@ -7,7 +7,6 @@
 using osu.Game.Configuration;
 using osuTK;
 using osu.Game.Rulesets.Judgements;
-using osu.Game.Rulesets.Objects.Drawables;
 using osu.Game.Rulesets.Scoring;
 using osu.Game.Skinning;
 using osuTK.Graphics;
@@ -19,12 +18,12 @@ public class DrawableOsuJudgement : DrawableJudgement
         private SkinnableSprite lighting;
         private Bindable<Color4> lightingColour;
 
-        public DrawableOsuJudgement(JudgementResult result, DrawableHitObject judgedObject)
-            : base(result, judgedObject)
+        public DrawableOsuJudgement()
         {
         }
 
-        public DrawableOsuJudgement()
+        public DrawableOsuJudgement(DrawableJudgementData model)
+            : base(model)
         {
         }
 
@@ -43,9 +42,9 @@ private void load(OsuConfigManager config)
             }
         }
 
-        protected override void PrepareForUse()
+        protected override void PrepareForUse(DrawableJudgementData model)
         {
-            base.PrepareForUse();
+            base.PrepareForUse(model);
 
             lightingColour?.UnbindAll();
 
diff --git a/osu.Game.Rulesets.Osu/UI/OsuPlayfield.cs b/osu.Game.Rulesets.Osu/UI/OsuPlayfield.cs
index 2eff99bd3..888c975bc 100644
--- a/osu.Game.Rulesets.Osu/UI/OsuPlayfield.cs
+++ b/osu.Game.Rulesets.Osu/UI/OsuPlayfield.cs
@@ -33,7 +33,7 @@ public class OsuPlayfield : Playfield
 
         protected override GameplayCursorContainer CreateCursor() => new OsuCursorContainer();
 
-        private readonly IDictionary<HitResult, DrawablePool<DrawableOsuJudgement>> poolDictionary = new Dictionary<HitResult, DrawablePool<DrawableOsuJudgement>>();
+        private readonly IDictionary<HitResult, DrawableJudgementPool> poolDictionary = new Dictionary<HitResult, DrawableJudgementPool>();
 
         public OsuPlayfield()
         {
@@ -109,14 +109,9 @@ private void onNewResult(DrawableHitObject judgedObject, JudgementResult result)
 
             var osuObject = (OsuHitObject)judgedObject.HitObject;
 
-            DrawableOsuJudgement explosion = poolDictionary[result.Type].Get(doj =>
-            {
-                doj.JudgedObject = judgedObject;
-
-                // todo: move to JudgedObject property?
-                doj.Position = osuObject.StackedEndPosition;
-                doj.Scale = new Vector2(osuObject.Scale);
-            });
+            DrawableOsuJudgement explosion = poolDictionary[result.Type].Get(new DrawableJudgementData(result, judgedObject));
+            explosion.Position = osuObject.StackedEndPosition;
+            explosion.Scale = new Vector2(osuObject.Scale);
 
             judgementLayer.Add(explosion);
         }
@@ -128,7 +123,7 @@ private class ApproachCircleProxyContainer : LifetimeManagementContainer
             public void Add(Drawable approachCircleProxy) => AddInternal(approachCircleProxy);
         }
 
-        private class DrawableJudgementPool : DrawablePool<DrawableOsuJudgement>
+        private class DrawableJudgementPool : DrawablePool<DrawableJudgementData, DrawableOsuJudgement>
         {
             private readonly HitResult result;
 
@@ -139,14 +134,7 @@ public DrawableJudgementPool(HitResult result)
             }
 
             protected override DrawableOsuJudgement CreateNewDrawable()
-            {
-                var judgement = base.CreateNewDrawable();
-
-                // just a placeholder to initialise the correct drawable hierarchy for this pool.
-                judgement.Result = new JudgementResult(new HitObject(), new Judgement()) { Type = result };
-
-                return judgement;
-            }
+                => new DrawableOsuJudgement(new DrawableJudgementData(new JudgementResult(new HitObject(), new Judgement()) { Type = result }, null));
         }
     }
 }
diff --git a/osu.Game/Rulesets/Judgements/DrawableJudgement.cs b/osu.Game/Rulesets/Judgements/DrawableJudgement.cs
index 3ec532629..957001884 100644
--- a/osu.Game/Rulesets/Judgements/DrawableJudgement.cs
+++ b/osu.Game/Rulesets/Judgements/DrawableJudgement.cs
@@ -1,10 +1,9 @@
 ﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
-using System.Diagnostics;
+using JetBrains.Annotations;
 using osuTK;
 using osu.Framework.Allocation;
-using osu.Framework.Caching;
 using osu.Framework.Extensions;
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Containers;
@@ -21,31 +20,18 @@ namespace osu.Game.Rulesets.Judgements
     /// <summary>
     /// A drawable object which visualises the hit result of a <see cref="Judgements.Judgement"/>.
     /// </summary>
-    public class DrawableJudgement : PoolableDrawable
+    public class DrawableJudgement : PoolableDrawable<DrawableJudgementData>
     {
         private const float judgement_size = 128;
 
         [Resolved]
         private OsuColour colours { get; set; }
 
-        private readonly Cached drawableCache = new Cached();
+        [CanBeNull]
+        public JudgementResult Result => Model.Result;
 
-        private JudgementResult result;
-
-        public JudgementResult Result
-        {
-            get => result;
-            set
-            {
-                if (result?.Type == value.Type)
-                    return;
-
-                result = value;
-                drawableCache.Invalidate();
-            }
-        }
-
-        public DrawableHitObject JudgedObject;
+        [CanBeNull]
+        public DrawableHitObject JudgedObject => Model.JudgedObject;
 
         protected Container JudgementBody;
         protected SpriteText JudgementText;
@@ -60,24 +46,34 @@ public JudgementResult Result
         /// </summary>
         protected virtual double FadeOutDelay => FadeInDuration;
 
+        protected DrawableJudgement()
+            : this(null, null)
+        {
+        }
+
         /// <summary>
         /// Creates a drawable which visualises a <see cref="Judgements.Judgement"/>.
         /// </summary>
         /// <param name="result">The judgement to visualise.</param>
         /// <param name="judgedObject">The object which was judged.</param>
-        public DrawableJudgement(JudgementResult result, DrawableHitObject judgedObject)
-            : this()
+        protected DrawableJudgement(JudgementResult result, DrawableHitObject judgedObject)
+            : this(new DrawableJudgementData(result, judgedObject))
         {
-            Result = result;
-            JudgedObject = judgedObject;
         }
 
-        public DrawableJudgement()
+        protected DrawableJudgement(DrawableJudgementData model)
+            : base(model)
         {
             Size = new Vector2(judgement_size);
             Origin = Anchor.Centre;
         }
 
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            prepareDrawables();
+        }
+
         protected virtual void ApplyHitAnimations()
         {
             JudgementBody.ScaleTo(0.9f);
@@ -86,27 +82,18 @@ protected virtual void ApplyHitAnimations()
             this.Delay(FadeOutDelay).FadeOut(400);
         }
 
-        [BackgroundDependencyLoader]
-        private void load()
-        {
-            prepareDrawables();
-        }
-
-        protected override void PrepareForUse()
+        protected override void PrepareForUse(DrawableJudgementData model)
         {
-            base.PrepareForUse();
+            base.PrepareForUse(model);
 
-            Debug.Assert(Result != null);
-
-            if (!drawableCache.IsValid)
-                prepareDrawables();
+            prepareDrawables();
 
             this.FadeInFromZero(FadeInDuration, Easing.OutQuint);
             JudgementBody.ScaleTo(1);
             JudgementBody.RotateTo(0);
             JudgementBody.MoveTo(Vector2.Zero);
 
-            switch (Result.Type)
+            switch (model.Result.Type)
             {
                 case HitResult.None:
                     break;
@@ -129,9 +116,14 @@ protected override void PrepareForUse()
             Expire(true);
         }
 
+        private HitResult? drawableType;
+
         private void prepareDrawables()
         {
-            var type = Result?.Type ?? HitResult.Perfect; //TODO: better default type from ruleset
+            var type = Model.Result?.Type ?? HitResult.Perfect; //TODO: better default type from ruleset
+
+            if (type == drawableType)
+                return;
 
             InternalChild = JudgementBody = new Container
             {
@@ -147,7 +139,22 @@ private void prepareDrawables()
                 }, confineMode: ConfineMode.NoScaling)
             };
 
-            drawableCache.Validate();
+            drawableType = type;
+        }
+    }
+
+    public readonly struct DrawableJudgementData
+    {
+        [NotNull]
+        public readonly JudgementResult Result;
+
+        [CanBeNull]
+        public readonly DrawableHitObject JudgedObject;
+
+        public DrawableJudgementData([NotNull] JudgementResult result, [CanBeNull] DrawableHitObject judgedObject)
+        {
+            Result = result;
+            JudgedObject = judgedObject;
         }
     }
 }
```

What's your thoughts on this?

@bdach I also want your opinion on this, please.

**Comment by peppy** [comment]
On an initial check, your version seems okay, although I'm not sure I'd choose it over the way I implemented (much of a muchness?). I wanted to avoid adding even more levels of abstraction because it's already getting pretty silly (HitObject -> DrawableHitObject -> Judgement -> DrawableJudgement -> JudgementResult) working on this change was quite hard to follow this chain.

As for the "local implementations", I did mention that when working on this. It is probably something we will want to centralise once we get a feel for how this works. At the end of the day though, the idea is to make pooling flexible and leave it mostly up to the user how to best optimise (to reduce allocs caused by changes applied to the pooled drawables).

**Comment by smoogipoo** [comment]
Admittedly when implementing the above I didn't think about `SkinnableDrawable`, that came as an afterthought and wouldn't work well with my structure...

**Comment by smoogipoo** [comment]
Would you be fine with the following?:

```diff
diff --git a/osu.Game.Rulesets.Osu/UI/OsuPlayfield.cs b/osu.Game.Rulesets.Osu/UI/OsuPlayfield.cs
index 2eff99bd3..291a572fd 100644
--- a/osu.Game.Rulesets.Osu/UI/OsuPlayfield.cs
+++ b/osu.Game.Rulesets.Osu/UI/OsuPlayfield.cs
@@ -111,7 +111,7 @@ private void onNewResult(DrawableHitObject judgedObject, JudgementResult result)
 
             DrawableOsuJudgement explosion = poolDictionary[result.Type].Get(doj =>
             {
-                doj.JudgedObject = judgedObject;
+                doj.Apply(result, judgedObject);
 
                 // todo: move to JudgedObject property?
                 doj.Position = osuObject.StackedEndPosition;
@@ -143,7 +143,7 @@ protected override DrawableOsuJudgement CreateNewDrawable()
                 var judgement = base.CreateNewDrawable();
 
                 // just a placeholder to initialise the correct drawable hierarchy for this pool.
-                judgement.Result = new JudgementResult(new HitObject(), new Judgement()) { Type = result };
+                judgement.Apply(new JudgementResult(new HitObject(), new Judgement()) { Type = result }, null);
 
                 return judgement;
             }
diff --git a/osu.Game/Rulesets/Judgements/DrawableJudgement.cs b/osu.Game/Rulesets/Judgements/DrawableJudgement.cs
index 3ec532629..42514bc1e 100644
--- a/osu.Game/Rulesets/Judgements/DrawableJudgement.cs
+++ b/osu.Game/Rulesets/Judgements/DrawableJudgement.cs
@@ -2,9 +2,9 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using System.Diagnostics;
+using JetBrains.Annotations;
 using osuTK;
 using osu.Framework.Allocation;
-using osu.Framework.Caching;
 using osu.Framework.Extensions;
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Containers;
@@ -28,24 +28,8 @@ public class DrawableJudgement : PoolableDrawable
         [Resolved]
         private OsuColour colours { get; set; }
 
-        private readonly Cached drawableCache = new Cached();
-
-        private JudgementResult result;
-
-        public JudgementResult Result
-        {
-            get => result;
-            set
-            {
-                if (result?.Type == value.Type)
-                    return;
-
-                result = value;
-                drawableCache.Invalidate();
-            }
-        }
-
-        public DrawableHitObject JudgedObject;
+        public JudgementResult Result { get; private set; }
+        public DrawableHitObject JudgedObject { get; private set; }
 
         protected Container JudgementBody;
         protected SpriteText JudgementText;
@@ -68,8 +52,7 @@ public JudgementResult Result
         public DrawableJudgement(JudgementResult result, DrawableHitObject judgedObject)
             : this()
         {
-            Result = result;
-            JudgedObject = judgedObject;
+            Apply(result, judgedObject);
         }
 
         public DrawableJudgement()
@@ -78,6 +61,12 @@ public DrawableJudgement()
             Origin = Anchor.Centre;
         }
 
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            prepareDrawables();
+        }
+
         protected virtual void ApplyHitAnimations()
         {
             JudgementBody.ScaleTo(0.9f);
@@ -86,10 +75,10 @@ protected virtual void ApplyHitAnimations()
             this.Delay(FadeOutDelay).FadeOut(400);
         }
 
-        [BackgroundDependencyLoader]
-        private void load()
+        public void Apply([System.Diagnostics.CodeAnalysis.NotNull] JudgementResult result, [CanBeNull] DrawableHitObject judgedObject)
         {
-            prepareDrawables();
+            Result = result;
+            JudgedObject = judgedObject;
         }
 
         protected override void PrepareForUse()
@@ -98,8 +87,7 @@ protected override void PrepareForUse()
 
             Debug.Assert(Result != null);
 
-            if (!drawableCache.IsValid)
-                prepareDrawables();
+            prepareDrawables();
 
             this.FadeInFromZero(FadeInDuration, Easing.OutQuint);
             JudgementBody.ScaleTo(1);
@@ -129,10 +117,15 @@ protected override void PrepareForUse()
             Expire(true);
         }
 
+        private HitResult? currentDrawableType;
+
         private void prepareDrawables()
         {
             var type = Result?.Type ?? HitResult.Perfect; //TODO: better default type from ruleset
 
+            if (type == currentDrawableType)
+                return;
+
             InternalChild = JudgementBody = new Container
             {
                 Anchor = Anchor.Centre,
@@ -147,7 +140,7 @@ private void prepareDrawables()
                 }, confineMode: ConfineMode.NoScaling)
             };
 
-            drawableCache.Validate();
+            currentDrawableType = type;
         }
     }
 }
```
Main purpose is to reduce the uncertainty of how to initialise judgements. DHO follows a similar structure already - the `apply()` method was intended to be made public when pooling is implemented.

**Comment by peppy** [comment]
Not sure about that being the final iteration, but seems like a step further in the correct direction.

**Comment by bdach** [comment]
I haven't looked at this PR properly until today, as I was mostly waiting for the framework-side change to go in. There I just mostly looked at it and went "yep, it's a drawable pool" and figured it was sane based off the test cases, which was maybe a little naive. I did look at this today but not too thoroughly and my focus is off so this might not be the most useful opinion ever.

I do agree with the assessment that the PR in its initial state could be improved upon as the drawable preparation logic was a little scattered, as @smoogipoo said. The [first proposal](https://github.com/ppy/osu/pull/9431#issuecomment-656074008) I like for how clean it is in this case, but in general I don't think I necessarily agree with tying pooled drawables to a model class, which might not be all that valid always (thinking of something like particle effects - not much reason to force creation/passing of models for something that isn't really physical representation of data). While we could have pools for model-less drawables and pools for model-backed drawables that just feels like enterprise-grade design™ so probably not a good idea.

The [second patch](https://github.com/ppy/osu/pull/9431#issuecomment-656112874) I would wholeheartedly agree with and while it doesn't iron out all of the kinks it's definitely better. Although maybe I'd expand on that version in the following manner:

<details>
<summary>patch proposing moving out further setup logic from <code>Get()</code> to a now-virtual <code>Apply()</code></summary>

```diff
diff --git a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuJudgement.cs b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuJudgement.cs
index 9d0c40629..fa980c758 100644
--- a/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuJudgement.cs
+++ b/osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuJudgement.cs
@@ -43,6 +43,17 @@ private void load(OsuConfigManager config)
             }
         }
 
+        public override void Apply(JudgementResult result, DrawableHitObject judgedObject)
+        {
+            base.Apply(result, judgedObject);
+
+            if (judgedObject?.HitObject is OsuHitObject osuObject)
+            {
+                Position = osuObject.StackedPosition;
+                Scale = new Vector2(osuObject.Scale);
+            }
+        }
+
         protected override void PrepareForUse()
         {
             base.PrepareForUse();
diff --git a/osu.Game.Rulesets.Osu/UI/OsuPlayfield.cs b/osu.Game.Rulesets.Osu/UI/OsuPlayfield.cs
index 291a572fd..474e7c0f9 100644
--- a/osu.Game.Rulesets.Osu/UI/OsuPlayfield.cs
+++ b/osu.Game.Rulesets.Osu/UI/OsuPlayfield.cs
@@ -107,16 +107,7 @@ private void onNewResult(DrawableHitObject judgedObject, JudgementResult result)
             if (!judgedObject.DisplayResult || !DisplayJudgements.Value)
                 return;
 
-            var osuObject = (OsuHitObject)judgedObject.HitObject;
-
-            DrawableOsuJudgement explosion = poolDictionary[result.Type].Get(doj =>
-            {
-                doj.Apply(result, judgedObject);
-
-                // todo: move to JudgedObject property?
-                doj.Position = osuObject.StackedEndPosition;
-                doj.Scale = new Vector2(osuObject.Scale);
-            });
+            DrawableOsuJudgement explosion = poolDictionary[result.Type].Get(doj => doj.Apply(result, judgedObject));
 
             judgementLayer.Add(explosion);
         }
diff --git a/osu.Game/Rulesets/Judgements/DrawableJudgement.cs b/osu.Game/Rulesets/Judgements/DrawableJudgement.cs
index 1671e97db..4e7f0018e 100644
--- a/osu.Game/Rulesets/Judgements/DrawableJudgement.cs
+++ b/osu.Game/Rulesets/Judgements/DrawableJudgement.cs
@@ -75,7 +75,7 @@ protected virtual void ApplyHitAnimations()
             this.Delay(FadeOutDelay).FadeOut(400);
         }
 
-        public void Apply([NotNull] JudgementResult result, [CanBeNull] DrawableHitObject judgedObject)
+        public virtual void Apply([NotNull] JudgementResult result, [CanBeNull] DrawableHitObject judgedObject)
         {
             Result = result;
             JudgedObject = judgedObject;
```

</details>

File: osu.Game/Rulesets/Judgements/DrawableJudgement.cs [hunk 1]
```diff
@@ -81,11 +86,25 @@ protected virtual void ApplyHitAnimations()
...
+
+            Debug.Assert(Result != null);
+
+            if (!drawableCache.IsValid)
+                prepareDrawables();
 
             this.FadeInFromZero(FadeInDuration, Easing.OutQuint);
+            JudgementBody.ScaleTo(1);
+            JudgementBody.RotateTo(0);
+            JudgementBody.MoveTo(Vector2.Zero);
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game/Rulesets/Judgements/DrawableJudgement.cs:107]: hmm, so far we've always done the resets in the `Apply` methods (see `UpdateState` of `DrawableHitObjects`) so I went with that here too.

[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game/Rulesets/Judgements/DrawableJudgement.cs:107]: I agree with this being here for the same reason.

[the remark this answers is not in the document]
Comment (bdach) [reply osu.Game/Rulesets/Judgements/DrawableJudgement.cs:107]: OK, I was not aware there was precedent for this.

**Comment by peppy** [comment]
Looks good from my end.

**Comment by peppy** [comment]
@smoogipoo wanna double-check this one? should just be as-above proposed.

File: osu.Game/Rulesets/Judgements/DrawableJudgement.cs [hunk 2]
```diff
@@ -48,29 +50,21 @@ public class DrawableJudgement : CompositeDrawable
...
         {
-            Result = result;
-            JudgedObject = judgedObject;
+            Apply(result, judgedObject);
+        }
 
+        public DrawableJudgement()
+        {
             Size = new Vector2(judgement_size);
+            Origin = Anchor.Centre;
```
Comment (smoogipoo) [inline osu.Game/Rulesets/Judgements/DrawableJudgement.cs:61]: Is this required? Seems to break taiko judgement positioning (and maybe mania, depends on how it positions it).

Can it be moved to osu! instead if it's required?

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/Judgements/DrawableJudgement.cs (hunk 2 above)
Comment (peppy) [reply osu.Game/Rulesets/Judgements/DrawableJudgement.cs:61]: Interesting, I didn't even notice. I just moved it here because I thought it was shared across ruleset, it can be reverted without issue.

Comment (smoogipoo) [reply osu.Game/Rulesets/Judgements/DrawableJudgement.cs:61]: Actually doesn't seem to be as a result of this PR and centre origin seems to be the general use case (all our rulesets + tau + sentakki)

**Review by smoogipoo — APPROVED** [review]
